crates/sandbox/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+use std::collections::HashMap;
+
+use config::DevnetConfig;
+use port_binding::PortBinding;
+
+pub mod config;
+pub mod protocol;
+
+#[derive(Debug)]
+pub struct Devnet {
+    config: DevnetConfig,
+    pub nodes: HashMap<String, Node>,
+}
+
+#[derive(Debug)]
+pub struct Node {
+    pub name: String,
+    pub swarm_addr: String,
+    pub server_addr: String,
+}
+
+impl Devnet {
+    pub fn new(config: DevnetConfig) -> Self {
+        Self {
+            config,
+            nodes: HashMap::new(),
+        }
+    }
+
+    pub async fn start(&mut self) -> eyre::Result<()> {
+        self.start_nodes().await?;
+        self.print_info();
+        Ok(())
+    }
+
+    async fn start_nodes(&mut self) -> eyre::Result<()> {
+        let mut swarm_port = self.config.start_swarm_port;
+        let mut server_port = self.config.start_server_port;
+
+        for i in 0..self.config.node_count {
+            let node_name = format!("node{}", i + 1);
+
+            let swarm_binding =
+                PortBinding::next_available(self.config.swarm_host.parse()?, &mut swarm_port)
+                    .await?;
+            let swarm_port_num = swarm_binding.port();
+
+            let server_binding =
+                PortBinding::next_available(self.config.server_host.parse()?, &mut server_port)
+                    .await?;
+            let server_port_num = server_binding.port();
+
+            let node = Node {
+                name: node_name.clone(),
+                swarm_addr: format!("{}:{}", self.config.swarm_host, swarm_port_num),
+                server_addr: format!("{}:{}", self.config.server_host, server_port_num),
+            };
+
+            self.nodes.insert(node_name, node);
+        }
+
+        Ok(())
+    }
+
+    fn print_info(&self) {
+        println!("Devnet running with {} nodes:", self.nodes.len());
+
+        for node in self.nodes.values() {
+            println!("\nNode: {}", node.name);
+            println!("  Swarm: {}", node.swarm_addr);
+            println!("  RPC: {}", node.server_addr);
+        }
+
+        println!("\nProtocols enabled: {}", self.config.protocols.join(","));
+    }
+}
+
+pub mod port_binding {
+    use std::net::{IpAddr, SocketAddr};
+
+    use eyre::bail;
+    use tokio::net::TcpListener;
+
+    pub struct PortBinding {
+        address: SocketAddr,
```

Comment:
you didn't also migrate the tests from `e2e-tests`

https://github.com/calimero-network/core/blob/ad54bb5908afb0866d09920f5c33be4041d676d0/e2e-tests/src/driver.rs#L643-L705

and consider moving this to it's own file if you want it to be in it's own module